mava/networks.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+from typing import Dict, Sequence, Tuple, Union
+
+import chex
+import distrax
+import jax
+import jax.numpy as jnp
+import numpy as np
+from flax import linen as nn
+from flax.linen.initializers import constant, orthogonal
+
+from mava.types import (
+    Observation,
+    ObservationGlobalState,
+    RNNGlobalObservation,
+    RNNObservation,
+)
+
+
+class MLPTorso(nn.Module):
+    """MLP torso."""
+
+    layer_sizes: Sequence[int]
+    activation: str = "relu"
+    use_layer_norm: bool = False
+
+    def setup(self) -> None:
+        """Set up the activation function."""
+        if self.activation == "relu":
+            self.activation_fn = nn.relu
+        elif self.activation == "tanh":
+            self.activation_fn = nn.tanh
+
+    @nn.compact
+    def __call__(self, observation: chex.Array) -> chex.Array:
+        """Forward pass."""
+        x = observation
+        for layer_size in self.layer_sizes:
+            x = nn.Dense(
+                layer_size,
+                kernel_init=orthogonal(np.sqrt(2)),
+                bias_init=constant(0.0),
+            )(x)
+            if self.use_layer_norm:
+                x = nn.LayerNorm(use_scale=False)(x)
+            x = self.activation_fn(x)
+        return x
+
+
+class FeedForwardActor(nn.Module):
+    """Feedforward Actor Network."""
+
+    torso: nn.Module
+    num_actions: Sequence[int]
+
+    @nn.compact
+    def __call__(self, observation: Observation) -> distrax.Categorical:
+        """Forward pass."""
+        x = observation.agents_view
+
+        x = self.torso(x)
+        actor_output = nn.Dense(
+            self.num_actions, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(x)
```

Comment:
Had a quick look and by default bias_init is zeros, so *I think* we don't need that part. While we're here though why do we do the `kernel_init`? I assume the default wouldn't make a huge difference?

```suggestion
        actor_output = nn.Dense(self.num_actions, kernel_init=orthogonal(0.01))(x)
```

Reply:
I just moved the current network that we have without any investigation related to the net setup, I prefer we keep it with orthogonal initialization for now and investigate it because it is important for stability based on what I know and it was mentioned in the literature of PPO. Maybe we can make it as a feature :woman_shrugging:

Reply:
Cool, happy to keep the init, was just wondering if you knew why we did it